### PR TITLE
Unify `setup_expt.py` and `setup_expt_fcstonly.py`

### LINKED ISSUE #536: Unify `setup_expt.py` and `setup_expt_fcstonly.py`
In the present state, there are 2 entry points to the workflow setup.
One for `cycled` and another for `forecast-only` modes of operation.

As more "modes" are added, this adds duplication of code fragments and burden to the maintainers.

This issue documents the unification of `setup_expt.py` and `setup_expt_fcstonly.py`

### PR DESCRIPTION
**Pull request by aerorahul**
**Issues resolved**
closes #536 

**Description**
In this PR, `setup_expt.py` and `setup_expt_fcstonly.py` are unified under the former name.

The user is required to provide a `mode` as the first positional argument to `setup.py`.  Valid options are `cycled` and `forecast-only`.  

```
❯ python3 setup_expt.py -h
usage: setup_expt.py [-h] {cycled,forecast-only} ...

Setup files and directories to start a GFS parallel. Create EXPDIR, copy config files. Create COMROT experiment directory structure, link
initial condition files from $ICSDIR to $COMROT

positional arguments:
  {cycled,forecast-only}
    cycled              arguments for cycled mode
    forecast-only       arguments for forecast-only mode

optional arguments:
  -h, --help            show this help message and exit
```

Upon choosing one of these modes, options specific to the mode can be realized as follows for the `forecast-only` and `cycled` modes respectively.

```
❯ python3 setup_expt.py forecast-only -h                                                                  feature/unify-setups
usage: setup_expt.py forecast-only [-h] [--pslot PSLOT] [--resdet RESDET] [--comrot COMROT] [--expdir EXPDIR] --idate IDATE --edate EDATE
                                   [--icsdir ICSDIR] [--configdir CONFIGDIR] [--cdump CDUMP] [--gfs_cyc {0,1,2,4}] [--start {warm,cold}]
                                   [--app {ATM,ATMW,S2S,S2SW}]

optional arguments:
  -h, --help            show this help message and exit
  --pslot PSLOT         parallel experiment name
  --resdet RESDET       resolution of the deterministic model forecast
  --comrot COMROT       full path to COMROT
  --expdir EXPDIR       full path to EXPDIR
  --idate IDATE         starting date of experiment, initial conditions must exist!
  --edate EDATE         end date experiment
  --icsdir ICSDIR       full path to initial condition directory
  --configdir CONFIGDIR
                        full path to directory containing the config files
  --cdump CDUMP         CDUMP to start the experiment
  --gfs_cyc {0,1,2,4}   GFS cycles to run
  --start {warm,cold}   restart mode: warm or cold
  --app {ATM,ATMW,S2S,S2SW}
                        UFS application
```

```
❯ python3 setup_expt.py cycled -h                                                                         feature/unify-setups
usage: setup_expt.py cycled [-h] [--pslot PSLOT] [--resdet RESDET] [--comrot COMROT] [--expdir EXPDIR] --idate IDATE --edate EDATE
                            [--icsdir ICSDIR] [--configdir CONFIGDIR] [--cdump CDUMP] [--gfs_cyc {0,1,2,4}] [--start {warm,cold}]
                            [--resens RESENS] [--nens NENS] [--app {ATM,ATMW}]

optional arguments:
  -h, --help            show this help message and exit
  --pslot PSLOT         parallel experiment name
  --resdet RESDET       resolution of the deterministic model forecast
  --comrot COMROT       full path to COMROT
  --expdir EXPDIR       full path to EXPDIR
  --idate IDATE         starting date of experiment, initial conditions must exist!
  --edate EDATE         end date experiment
  --icsdir ICSDIR       full path to initial condition directory
  --configdir CONFIGDIR
                        full path to directory containing the config files
  --cdump CDUMP         CDUMP to start the experiment
  --gfs_cyc {0,1,2,4}   GFS cycles to run
  --start {warm,cold}   restart mode: warm or cold
  --resens RESENS       resolution of the ensemble model forecast
  --nens NENS           number of ensemble members
  --app {ATM,ATMW}      UFS application
```

Note, `cycled` mode presents some extra options e.g. `nens` as well as a reduced list of the UFS weather model applications.

There are some irrelevant sections such as `gfs_cyc` in forecast-only that is still preserved in this PR.   It will be cleaned up in subsequent PR's.
Another unnecessary complication is the argument of `--start`.  The logic presented here would ideally be selected at runtime based on the type of IC's populated in comrot.  It is left unchanged.

This PR will no doubt need an update on the Wiki and any other relevant documentation.  
Those updates should be performed after this PR has been reviewed and tested, but prior to merge into `develop`.

**Impact Statement** 
No impact is expected as a result of these changes in the execution of the experiment.  

**Testing Status**
In progress

### CODE REVIEW

**Comment by aerorahul** [comment]
@WalterKolczynski-NOAA @JessicaMeixner-NOAA @KateFriedman-NOAA 
I have marked the PR as ready for review (since Draft does not get much attention), but it may contain somethings that might not be to your liking.
Please review and I can work on resolving your feedback and comments.

File: ush/rocoto/setup_expt.py
Comment (WalterKolczynski-NOAA) [inline ush/rocoto/setup_expt.py:15]: Instead of this, why not use os.makedirs directly with exist_ok=True? While I'm not really worried about the race condition introduced by this construct, I do like using stock functions when they do the same thing better.

File: ush/rocoto/setup_expt.py
Comment (WalterKolczynski-NOAA) [inline ush/rocoto/setup_expt.py:23]: Technically this should have a docblock, but the existing functions don't necessarily have one either.

File: ush/rocoto/setup_expt.py
Comment (WalterKolczynski-NOAA) [inline ush/rocoto/setup_expt.py]: If we are going to update the loop variable, I'd prefer it be to something that indicates what it represents, like 'mem'.

File: ush/rocoto/setup_expt.py
Comment (WalterKolczynski-NOAA) [inline ush/rocoto/setup_expt.py:158]: This is a much better construction

File: ush/rocoto/setup_expt.py
Comment (WalterKolczynski-NOAA) [inline ush/rocoto/setup_expt.py]: Thank you for adding the default. Almost no one should need to actually set this.

**Review by WalterKolczynski-NOAA — DISMISSED** [review]
Overall a big improvement. I've noted a couple of things I'd like to see changed, but none are something I would insist on.

**Comment by JessicaMeixner-NOAA** [comment]
@aerorahul seems like a great thing to unify these to scripts.  I'd recommend with this PR adding usage information to the README https://github.com/NOAA-EMC/global-workflow/blob/feature/unify-setup-expt/README.md or make sure there's corresponding usage info updated on the wiki.  

As I have not done much with the set-up scripts if anything at all, I'm not sure I'm a great choice for a reviewer of this PR, but could test to make sure the coupled free-forecast runs if someone else is not already doing that test.

File: ush/rocoto/setup_expt.py
Comment (aerorahul) [reply ush/rocoto/setup_expt.py:15]: I can make the change you suggest.  I was simply changing the input argument from `d` to `dirname`.
In general, single character arguments or variables are hard to grep and non-descriptive.

File: ush/rocoto/setup_expt.py
Comment (aerorahul) [reply ush/rocoto/setup_expt.py:23]: I will put some documentation.

File: ush/rocoto/setup_expt.py
Comment (aerorahul) [reply ush/rocoto/setup_expt.py]: Same argument applies to single character iterators.  I thought about `mem`, but then chose not to since `mem` is used as a string in this loop, and using `mem` as an iterator would be confusing.   I could have chosen `nn` instead.

**Comment by aerorahul** [comment]
> @aerorahul seems like a great thing to unify these to scripts. I'd recommend with this PR adding usage information to the README https://github.com/NOAA-EMC/global-workflow/blob/feature/unify-setup-expt/README.md or make sure there's corresponding usage info updated on the wiki.
> 
> As I have not done much with the set-up scripts if anything at all, I'm not sure I'm a great choice for a reviewer of this PR, but could test to make sure the coupled free-forecast runs if someone else is not already doing that test.

I will update the README as you suggest.
Your help in testing the  individual`setup_expt.py` and `setup_expt_fcstonly.py` with the consolidated `setup_expt.py` for coupled setups would be appreciated.
I performed the setups on Orion and did a `diff -r` between the `develop` and the `pr` branch, and found largely no differences.  A more careful eye would be helpful.

File: ush/rocoto/setup_expt.py
Comment (aerorahul) [reply ush/rocoto/setup_expt.py:23]: done.

**Comment by aerorahul** [comment]
> > @aerorahul seems like a great thing to unify these to scripts. I'd recommend with this PR adding usage information to the README https://github.com/NOAA-EMC/global-workflow/blob/feature/unify-setup-expt/README.md or make sure there's corresponding usage info updated on the wiki.
> > As I have not done much with the set-up scripts if anything at all, I'm not sure I'm a great choice for a reviewer of this PR, but could test to make sure the coupled free-forecast runs if someone else is not already doing that test.
> 
> I will update the README as you suggest. Your help in testing the individual`setup_expt.py` and `setup_expt_fcstonly.py` with the consolidated `setup_expt.py` for coupled setups would be appreciated. I performed the setups on Orion and did a `diff -r` between the `develop` and the `pr` branch, and found largely no differences. A more careful eye would be helpful.

Since the current `README` does not have any `setup` information, and all of that information is contained in the Wiki, I would like to update the Wiki immediately after this PR is ready for merge.

**Review by WalterKolczynski-NOAA — APPROVED** [review]

**Comment by WalterKolczynski-NOAA** [comment]
Waiting on @KateFriedman-NOAA to review Tues before merging.

File: ush/rocoto/setup_expt.py
Comment (KateFriedman-NOAA) [inline ush/rocoto/setup_expt.py:231]: Wondering why the two --app arguments (cycled and forecasts) can't be combined into one? I see a different `choices` for both but they overlap. Is this a way of not allowing the `S2S` and `S2SW` apps in cycled mode? It's two places to maintain that list now.

File: ush/rocoto/workflow_utils.py
Comment (KateFriedman-NOAA) [inline ush/rocoto/workflow_utils.py:157]: Perhaps these machine blocks could be put in some kind of parm file(s) and these scripts could parse them? An idea for later, not for this PR.

**Review by KateFriedman-NOAA — APPROVED** [review]
THANK YOU FOR COMBINING THESE SCRIPTS! Overall I am very pleased with this effort and approve it. Have left one small comment and have one picky request: as someone who has and will continue to use the setup scripts a lot, can the `cycled` and `forecast-only` arguments be shortened, mainly the `forecast-only` one? Perhaps `free` or some other shorter form that is still clear as to what it is?

**Comment by JessicaMeixner-NOAA** [comment]
A few curious diffs when testing the coupled free-forecast set-ups: 

(see /work/noaa/marine/jmeixner/pr537 on orion, the directories are more/less self-explanatory likely, but ask if you need help finding something). 

There is no config.base.default in the exp directory.  (Not an issue, but a difference worth pointing out in case that was unintentional). 

The definition of the ICS directory: 
30,32c30,32
```
< 	<!ENTITY EXPDIR "/work/noaa/marine/jmeixner/pr537/dev/exp/dev01">
< 	<!ENTITY ROTDIR "/work/noaa/marine/jmeixner/pr537/dev/comrot/dev01">
< 	<!ENTITY ICSDIR "/work/noaa/marine/jmeixner/pr537/dev/comrot/FV3ICS">
---
> 	<!ENTITY EXPDIR "/work/noaa/marine/jmeixner/pr537/uni/exp/uni01">
> 	<!ENTITY ROTDIR "/work/noaa/marine/jmeixner/pr537/uni/comrot/uni01">
> 	<!ENTITY ICSDIR "/home/jmeixner">
```
I think this will be an issue because ICSDIR is used in the forecast job, but I don't understand why/how this changed.

**Review by JessicaMeixner-NOAA — CHANGES_REQUESTED** [review]
Why did the definition of ICSDIR change?  (Otherwise this is fine).

**Comment by JessicaMeixner-NOAA** [comment]
> Why did the definition of ICSDIR change? (Otherwise this is fine).

Or I should read the instructions better... but fair warning I'm not going to be the only person with the coupled group who makes this mistake.  Might be worth pointing out in the Wednesday tag-up.

**Comment by KateFriedman-NOAA** [comment]
Unless someone fixed it recently, I thought the ICSDIR feature was broken and needs fixing/rework. I may have missed something while in WCOSS2 land...

**Review by JessicaMeixner-NOAA — APPROVED** [review]
I understand the ICSDIR issue (user error), so this is good to go from what I can tell.  I just compared the exp directories, I did not run a full test.

**Comment by JessicaMeixner-NOAA** [comment]
> Unless someone fixed it recently, I thought the ICSDIR feature was broken and needs fixing/rework. I may have missed something while in WCOSS2 land...

The ICSDIR is used in jobs/rocoto/coupled_ic.sh so hopefully this isn't broken because this is needed to run coupled jobs.  Does someone need to run a coupled free forecast before this is merged?

**Comment by KateFriedman-NOAA** [comment]
> The ICSDIR is used in jobs/rocoto/coupled_ic.sh so hopefully this isn't broken because this is needed to run coupled jobs. Does someone need to run a coupled free forecast before this is merged?

Oh ok, ignore my comment then. Sounds like it's been updated and used now. :)

**Comment by aerorahul** [comment]
@KateFriedman-NOAA @WalterKolczynski-NOAA Please hold the merge.  I want to make sure the `ICSDIR` variable is correctly handled here.
In the past, `setup_expt.py` used to get the IC's for `cycled` for `SDATE` and populate them in `ROTDIR`.  That is the function that is currently broken as @KateFriedman-NOAA notes.
The `ICSDIR` has a different use case in the forecast only applications (including coupled).  I should be able to preserve that and fix the broken `cycled` issue.

**Review by JessicaMeixner-NOAA — CHANGES_REQUESTED** [review]
Okay, we do need to fix or update the ICSDIR before merging this.  I couldn't set a ICSDIR for the forecast only, so this likely is going to break the coupled model.   This variable is used in : https://github.com/NOAA-EMC/global-workflow/blob/develop/jobs/rocoto/coupled_ic.sh  and in ush/forecast_predet.sh and ush/forecast_postdet.sh

**Comment by WalterKolczynski-NOAA** [comment]
> Okay, we do need to fix or update the ICSDIR before merging this. I couldn't set a ICSDIR for the forecast only, so this likely is going to break the coupled model. This variable is used in : https://github.com/NOAA-EMC/global-workflow/blob/develop/jobs/rocoto/coupled_ic.sh and in ush/forecast_predet.sh and ush/forecast_postdet.sh

We checked on this during the CM meeting. For the current forecast-only script, the --icsdir argument is ignored and it is always the $COMROT/FV3ICS used for coupled. This clearly isn't correct. I think the plan is to leave it as this PR has it and make sure the change to requiring the --icsdir argument for coupled is well communicated in the announcement and the wiki, and I will mention it in the coupled meeting tomorrow as well.

**Comment by JessicaMeixner-NOAA** [comment]
@WalterKolczynski-NOAA can you please test a coupled case, because it does not get set as $COMROT/FV3ICS in coupled or I'm not understanding something.  This PR can stay as is, but something else has to change for coupled to work, unless I'm missing something but I don't think we want the forecasts ICs to be copied to:  ICSDIR ="/home/jmeixner" as I"m pretty sure my home directory is going to fill up super fast.

**Comment by aerorahul** [comment]
The basis of the confusion is that `ICSDIR` is the source of ICs for cycled and free atm forecasts (location from which files are copied from), while it is the destination for ICs for the coupled system (destination where files are copied to for staging).

One solution for the coupled forecasts is to provide `--icsdir` as the path to where they wish to copy their initial conditions to.  The default is `$HOME` since that is where the cycled and free atm forecasts are being copied from (the directory can be linked from there).

**Comment by WalterKolczynski-NOAA** [comment]
> @WalterKolczynski-NOAA can you please test a coupled case, because it does not get set as $COMROT/FV3ICS in coupled or I'm not understanding something. This PR can stay as is, but something else has to change for coupled to work, unless I'm missing something but I don't think we want the forecasts ICs to be copied to: ICSDIR ="/home/jmeixner" as I"m pretty sure my home directory is going to fill up super fast.

It works for me:
```
Orion-login-4 rocoto/>./setup_expt.py forecast-only --app S2SW --res 384 --comrot $NOSCRUB/global-workflow --icsdir $NOSCRUB/global-workflow/FV3ICS --expdir $SAVE/global-workflow/expdir --idate 2013040100 --edate 2013040100 --pslot pr537

EDITED:  /work/noaa/global/wkolczyn/save/global-workflow/expdir/config.base as per user input.
DEFAULT: /work/noaa/global/wkolczyn/save/global-workflow/expdir/config.base.emc.dyn is for reference only.
Please verify and delete the default file before proceeding.

Orion-login-4 rocoto/>cd $SAVE/global-workflow/expdir/pr537
Orion-login-4 pr537/>grep ICSDIR *
config.base:export ICSDIR="/work/noaa/global/wkolczyn/noscrub/global-workflow/FV3ICS"
config.base.emc.dyn:export ICSDIR="@ICSDIR@"
```

**Comment by JessicaMeixner-NOAA** [comment]
Okay, I see what I did wrong an error if I try to specify ICSDIR, I should have specified icsdir.  Apologies.

**Comment by WalterKolczynski-NOAA** [comment]
We've all done it

**Review by JessicaMeixner-NOAA — APPROVED** [review]
Apologies -- when I said it didn't work I had ICSDIR in all caps and not lowercase. Specifying icsdir will work.

File: ush/rocoto/setup_expt.py
Comment (DavidHuber-NOAA) [inline ush/rocoto/setup_expt.py]: A minor point.  This should remain `{expdir}` as `{inputs.expdir}` points to `$EXPDIR/..`.

Comment (WalterKolczynski-NOAA) [reply ush/rocoto/setup_expt.py]: I don't think expdir is defined in this namespace. {base_config} should be used anyway.

Comment (DavidHuber-NOAA) [reply ush/rocoto/setup_expt.py]: `expdir` is defined in the namespace since it is part of `__main__`.  `base_config` works for line 163, but not 164 (apologies for not highlighting 164 in the above snippet).  However, a similar assignment could be used for config.base.emc.dyn.

Comment (DavidHuber-NOAA) [reply ush/rocoto/setup_expt.py]: Ah, that works.

Comment (WalterKolczynski-NOAA) [reply ush/rocoto/setup_expt.py]: Should be fixed now.

**Review by DavidHuber-NOAA — APPROVED** [review]

**Review by WalterKolczynski-NOAA — APPROVED** [review]

**Pull request merged**
